habitat_baselines/rl/ppo/ppo_trainer.py

```diff
@@ -0,0 +1,738 @@
+                    stats_episodes[
+                        (
+                            current_episodes[i].scene_id,
+                            current_episodes[i].episode_id,
+                        )
+                    ] = episode_stats
+                    if args.video_option:
+                        generate_video(
+                            args,
+                            rgb_frames[i],
+                            current_episodes[i].episode_id,
+                            cur_ckpt_idx,
+                            infos[i]["spl"],
+                            writer,
+                        )
+                        rgb_frames[i] = []
+
+                # episode continues
+                elif args.video_option:
+                    frame = observations_to_image(observations[i], infos[i])
+                    rgb_frames[i].append(frame)
+
+            # pausing self.envs with no new episode
+            if len(envs_to_pause) > 0:
+                state_index = list(range(self.envs.num_envs))
+                for idx in reversed(envs_to_pause):
+                    state_index.pop(idx)
+                    self.envs.pause_at(idx)
+
+                # indexing along the batch dimensions
+                test_recurrent_hidden_states = test_recurrent_hidden_states[
+                    state_index
+                ]
+                not_done_masks = not_done_masks[state_index]
+                current_episode_reward = current_episode_reward[state_index]
+
+                for k, v in batch.items():
+                    batch[k] = v[state_index]
+
+                if args.video_option:
+                    rgb_frames = [rgb_frames[i] for i in state_index]
+
+        aggregated_stats = dict()
+        for stat_key in next(iter(stats_episodes.values())).keys():
+            aggregated_stats[stat_key] = sum(
+                [v[stat_key] for v in stats_episodes.values()]
+            )
+        num_episodes = len(stats_episodes)
+
+        episode_reward_mean = aggregated_stats["reward"] / num_episodes
+        episode_spl_mean = aggregated_stats["spl"] / num_episodes
+        episode_success_mean = aggregated_stats["success"] / num_episodes
+
+        logger.info(
+            "Average episode reward: {:.6f}".format(episode_reward_mean)
+        )
+        logger.info(
+            "Average episode success: {:.6f}".format(episode_success_mean)
+        )
+        logger.info("Average episode SPL: {:.6f}".format(episode_spl_mean))
+
+        writer.add_scalars(
+            "eval_reward",
+            {"average reward": episode_reward_mean},
+            cur_ckpt_idx,
+        )
+        writer.add_scalars(
+            "eval_SPL", {"average SPL": episode_spl_mean}, cur_ckpt_idx
+        )
+        writer.add_scalars(
+            "eval_success",
+            {"average success": episode_success_mean},
+            cur_ckpt_idx,
+        )
+
+
+def generate_video(
```

Comment:
Maybe, move to ppo_uitils or etc all other classes/functions except PPOTrainer.  PPOTrainer already large enough.